src/rewrite_gelu.cpp

```diff
@@ -53,7 +58,60 @@ struct find_gelu_erf
     }
 };
 
-void rewrite_gelu::apply(module& m) const { match::find_matches(m, find_gelu_erf{}); }
+/**
+ * Find fastGELU blocks (where the graph already does a GELU approximation) and replace them
+ * with an alternative approximation that is less likely to overflow.
+ * The replacement approximation is equivalent to:
+ * GELU(x) ~= 0.5 * x * ( 1 + tanh( sqrt(2/M_PI) * (x + 0.044715 * x^3)))
+ */
+struct find_tanh_fast_gelu
+{
+    auto matcher() const { return match::gelu_tanh(); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        // Algebraically rearranged version of the tanh GELU approximation that is more numerically
+        // stable
+        auto ins        = r.result;
+        auto x          = r.instructions["x"];
+        auto sqrt_2_rpi = m.add_literal(
+            literal{shape{x->get_shape().type()},
+                    {0.7978845608028653558798921198687637369517172623298693153318516593}});
+        auto fit_const = m.add_literal(literal{shape{x->get_shape().type()}, {0.044715f}});
+        auto one       = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
+        auto xb        = insert_common_op(m, ins, make_op("mul"), {x, sqrt_2_rpi});
+        auto a         = insert_common_op(m, ins, make_op("mul"), {xb, fit_const});
+        auto b         = m.insert_instruction(ins, make_op("mul"), a, x);
+        auto c         = m.insert_instruction(ins, make_op("mul"), b, x);
+        auto u         = m.insert_instruction(ins, make_op("add"), c, xb);
+        auto neg_u     = m.insert_instruction(ins, make_op("neg"), u);
```

Comment:
I don't understand how tanh is computed here, please. How are you calculating e**(x) and e**(-x) here? Thanks.

Reply:
[tanh_gelu_scan2.pdf](https://github.com/ROCm/AMDMIGraphX/files/13832504/tanh_gelu_scan2.pdf)
I've attached scanned document with the derivation of the rearranged approximation.